feat(core): trace data type

prognant opened the pull request:
Work for this [RFC](https://github.com/vectordotdev/vector/blob/master/rfcs/2021-10-15-9572-accept-datadog-traces.md):

TL;DR: 
```
 pub enum Event {
     Log(LogEvent),
     Metric(Metric),
+    Trace(LogEvent),
 }
```

Include graphql changes:
* Output types are represented by a list on API responses
* Output types can be queried with strict (in)equality and type inclusion/exclusion  
```
outputType: {
	equals: [LOG, METRIC]
	notEquals: [LOG, TRACE]
        contains: LOG
        notContains: TRACE
      }
```

I did some early test with dummy [transforms](https://gist.github.com/prognant/5896b7432b25b28fd256935b055ee67f)

jszwedko commented on src/api/schema/components/source.rs:
```diff
@@ -15,17 +15,19 @@ use crate::{
 #[derive(Debug, Enum, Eq, PartialEq, Copy, Clone, Ord, PartialOrd)]
 pub enum SourceOutputType {
-    Any,
+    Multiple,
```
Did you consider updating this one to also be a bitmask? Corollary: do we need two separate types to model source outputs / component inputs? Could it just be one type like `EventType`?

prognant replied:
This enum seems to be used only for the graphql API (I may lack knowledge on this piece) as a [filtering option](https://github.com/vectordotdev/vector/blob/master/src/api/schema/filter.rs#L50) (and filtering by output type is not implemented for anything else than source if I'm correct). 
That being said I agree this would make much more sens to rework the filter and support boolean combinator on this filter instead of having this new `Multiple` option. I'll look at that option.

jszwedko replied:
Ah, I see. Yeah, I think we could update the API to allow, rather than a single type, a list of types as input. Something like:

```
pub struct EqualityFilter<T: InputType + PartialEq + Eq> {
    pub equals: Option<T>,
    pub not_equals: Option<T>,
    pub in: Vec<T>,
    pub not_in: Vec<T>
}
```

(keeping `equals` and `not_equals` for compatibility)

I don't think GraphQL supports boolean combinators.

prognant replied:
I like the `in` / `not_in` semantic, I will give it a try. Anyway using a bitmask as the underlying structure for the GraphQL query param works. I also noticed we do have the `or` combinator (but not `and` at the moment, it should not by a problem to add it). With the current iteration of the PR, a working GraphQL could look like:

```graphql
query {
  sources(filter: {
    outputType:{supports:TRACE}
    or: {outputType:{supports:LOG}}
  })
  {
    edges {
      node {
        componentId
      }
    }
  }
}
```

leebenson replied:
>  I also noticed we do have the or combinator (but not `and` at the moment, it should not by a problem to add it)

To clarify - `and` is the behaviour when multiple filter conditions are included, so there's not an explicit operator for that reason.

jszwedko commented on src/transforms/log_to_trace.rs:
```diff
@@ -0,0 +1,36 @@
+pub struct LogToTrace;
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "log_to_trace")]
```
Do we intend for users to use this transform? I noticed it wasn't documented.

prognant replied:
That's a good point. I don't think this one will be very useful outside testing, the opposite, `trace_to_log` sounds more useful. After rethinking I think there is no point to add this `log_to_trace` transform, I will remove it.

jszwedko commented on website/cue/reference/data_model/schema.cue:
```diff
@@ -349,4 +349,41 @@ data_model: schema: {
 		}
 	}
+
+	trace: {
```
I might suggest we leave out this documentation until we have a source / sink that uses traces otherwise users may be confused about why no components generate or support sending traces.

prognant replied:
That's probably a safe option, I will save that for later.

jszwedko reviewed (commented):
Generally looks good! Left just a few comments below.

leebenson commented on lib/vector-core/src/event/mod.rs:
```diff
@@ -77,14 +78,19 @@ impl Event {
     ///
     /// # Panics
     ///
     /// This function panics if self is anything other than an `Event::Log`.
```
Should the last line of this comment mention `Event::Trace`?

leebenson commented on lib/vector-core/src/event/mod.rs:
```diff
@@ -96,7 +102,7 @@ impl Event {
     /// This function panics if self is anything other than an `Event::Log`.
```
Nit: Comment is missing `Event::Trace`

leebenson commented on lib/vector-core/src/event/mod.rs:
```diff
@@ -108,7 +114,7 @@ impl Event {
     /// This function panics if self is anything other than an `Event::Log`.
```
Nit: Same here.

leebenson commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -321,23 +331,22 @@ impl From<Event> for VrlTarget {
 //   * If an element is anything else, assign to the `message` key.
 // * If `.` is anything else, assign to the `message` key.
-fn value_into_log_events(value: Value, metadata: EventMetadata) -> impl Iterator<Item = Event> {
+fn value_into_logevent(value: Value, metadata: EventMetadata) -> impl Iterator<Item = LogEvent> {
```
Was this renaming intentional? It implies singular, but returns an iterator of `LogEvent`.

prognant replied:
It was indeed intentional as an attempt to make the new return type, `LogEvent` instead of `Event` explicit, though I should have kept the plural.

leebenson commented on src/api/schema/components/source.rs:
```diff
@@ -55,8 +45,21 @@ impl Source {
     }
-    pub fn get_output_type(&self) -> SourceOutputType {
-        self.0.output_type.into()
+    pub fn get_output_types(&self) -> Vec<SourceOutputType> {
+        let mut t = vec![];
+        self.0
+            .output_type
+            .contains(DataType::Log)
+            .then(|| t.push(SourceOutputType::Log));
+        self.0
+            .output_type
+            .contains(DataType::Metric)
+            .then(|| t.push(SourceOutputType::Metric));
+        self.0
+            .output_type
+            .contains(DataType::Trace)
+            .then(|| t.push(SourceOutputType::Trace));
+        t
     }
```
Instead of the repeated logic here, I think this would be a little neater to implement a translation between `DataType` and `SourceOutputType`, and then iterate over the arms:

```rust
impl From<&SourceOutputType> for DataType {
    fn from(s: &SourceOutputType) -> Self {
        match s {
            SourceOutputType::Log => DataType::Log,
            SourceOutputType::Metric => DataType::Metric,
            SourceOutputType::Trace => DataType::Trace,
        }
    }
}

pub fn get_output_types(&self) -> Vec<SourceOutputType> {
    [
        SourceOutputType::Log,
        SourceOutputType::Metric,
        SourceOutputType::Trace,
    ]
    .into_iter()
    .filter(|s| self.0.output_type.contains(s.into()))
    .map(Into::into)
    .collect()
}
```

To save enumerating the options, you could use [strum](https://crates.io/crates/strum), which is used in grok parsing and is a dep of other lib we use.

leebenson replied:
FWIW, I tried the strum implementation just now - adding an [EnumIter](https://docs.rs/strum_macros/0.23.1/strum_macros/derive.EnumIter.html) to `SourceOutputType`:

```rust
#[derive(Debug, Enum, Eq, PartialEq, Copy, Clone, Ord, PartialOrd, EnumIter)]
pub enum SourceOutputType {
    Log,
    Metric,
    Trace,
}
```

And then the method becomes:

```rust
pub fn get_output_types(&self) -> Vec<SourceOutputType> {
    SourceOutputType::iter()
        .filter(|s| self.0.output_type.contains(s.into()))
        .collect()
}
```

leebenson commented on src/api/schema/components/source.rs:
```diff
@@ -90,8 +92,8 @@ impl Source {
     /// Source output type
-    pub async fn output_type(&self) -> SourceOutputType {
-        self.get_output_type()
+    pub async fn output_type(&self) -> Vec<SourceOutputType> {
```
Since this now returns an array, this should probably be renamed to `output_types`, with the doc comment updated to reflect.

This is a breaking change in the GraphQL API, but returning an array already breaks it, so this is somewhat equivalent.

leebenson commented on src/api/schema/components/source.rs:
```diff
@@ -69,10 +72,9 @@ impl sort::SortableByField<SourcesSortFieldName> for Source {
-                &SourceOutputType::from(rhs.0.output_type),
-            ),
+            SourcesSortFieldName::OutputType => {
+                Ord::cmp(&u8::from(self.0.output_type), &u8::from(rhs.0.output_type))
```
👍🏻 It might be worth adding a comment to `SourceOutputType` to keep its options alphabetical (which they already are), so that bitmask values match the intuitive sort order users likely expect.

Just in case this gets added to in the future and that connection isn't obvious.

prognant replied:
Indeed, the fact that the enum order drives how sorting happens definitely needs to be documented.

leebenson commented on src/api/schema/filter.rs:
```diff
@@ -46,8 +47,38 @@ impl StringFilter {
+    pub not_contains: Option<T>,
+}
+
+impl<T: InputType + PartialEq + Eq + std::cmp::Ord> ListFilter<T> {
```
Nit- `Ord` is already in scope:

```suggestion
impl<T: InputType + PartialEq + Eq + Ord> ListFilter<T> {
```

leebenson reviewed (changes requested):
Great work @prognant. I really like that you got under the hood of the GraphQL implementation, and augmented filtering -- nicely done!

A few minor comments to tighten up GraphQL field names, but nothing major is standing out at me.

I'd wait for @lukesteensen's input on the topology parts.

leebenson commented on src/api/schema/filter.rs:
```diff
@@ -46,8 +47,38 @@ impl StringFilter {
+}
+
+#[derive(InputObject)]
+#[graphql(concrete(name = "ComponentKindFilter", params(ComponentKind)))]
```
Nit: Since you introduced the `source::SourceOutputType` pattern, you could change this param to `source::ComponentKind` and drop the explicit import.

I like this better- it's clearer that it's being implemented on source components rather than having to check the import list for the scope.

prognant commented:
Thanks a lot for the review @leebenson 👍  I think I addressed all your suggestion in the last update.

leebenson reviewed (changes requested):
Just needs a `make build-graphql-schema` to pick up the latest schema changes.

lukesteensen commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -24,6 +24,7 @@ pub enum VrlTarget {
     // that `fields` must always be a `Map` variant.
     LogEvent(Value, EventMetadata),
     Metric(Metric),
+    Trace(Value, EventMetadata),
```
I'd be tempted to leave traces out of VRL for now, mostly because I'm not sure if we'll want to give them the same API as logs or do something more specific. One of the goals with the initial implementation is to limit the surface area so that it's easier to change as we get more experience with traces.

prognant replied:
I see one issue there, one of the major use case for traces is demultiplexing a trace flow based on various criterion, thus using the `filter`/`route` transforms, but we would need to be able to evaluate VRL condition. I don't see that as a blocker, as the whole VRL with trace thing need to be carefully though. So I would say it sounds reasonable, at the moment, to let traces out of VRL and plan that a bit later to gain a wider consensus.

JeanMertz replied:
Agreed on this. I'd rather we write an RFC first to think through the interaction model. We already have half-baked support for metrics that I'd wish we had left out until we're ready to fully commit to it, this seems like more baggage we're not ready to carry yet, until some thinking has been done on this.

prognant replied:
I see several alternatives:
1. We can keep VRL support for trace, but disable it at the topology level for the `remap` transform, letting only the VRL condition for `filter`/`route` transforms and preventing trace manipulation in VRL
2. We let VRL enabled including the `remap` transform
3. We let traces completely out of VRL

Users will probably expect to get at least `filter` & `route` to work, I'm not so sure about the rest.

lukesteensen commented on lib/vector-core/src/event/lua/event.rs:
```diff
@@ -9,6 +9,7 @@ impl<'a> ToLua<'a> for Event {
         match self {
             Event::Log(log) => table.raw_set("log", log.to_lua(lua)?)?,
             Event::Metric(metric) => table.raw_set("metric", metric.to_lua(lua)?)?,
+            Event::Trace(trace) => table.raw_set("trace", trace.to_lua(lua)?)?,
```
Similar to VRL, I'm not sure that we need to support processing traces via lua right off the bat. Might be simpler to defer this until we have a use case and think a bit more about the API design.

prognant replied:
I think letting LUA on the side is probably a good idea, as opposed to VRL, I'm not sure this should be a priority. We could probably wait for user demand.

lukesteensen commented on src/config/graph.rs:
```diff
@@ -176,10 +176,10 @@ impl Graph {
                 errors.push(format!(
-                    "Data type mismatch between {} ({:?}) and {} ({:?})",
-                    edge.from, from_ty, edge.to, to_ty
+                    "Data type mismatch between {} and {}",
+                    edge.from, edge.to
```
Is there not a good way to keep the types in this error message? I get that it's a bit more complex now, but it does seem like helpful information that'd be nice to keep.

lukesteensen reviewed (commented):
Looks like a good start! I imagine there's probably also a docs component to this where we'll want to introduce the new type there and update the inputs/outputs for the relevant components.

bruceg reviewed (commented):
I don't have much to add on this except to echo @lukesteensen's concerns. The implementation looks reasonable otherwise.

leebenson reviewed (approved):
The GraphQL changes look good 👍🏻 

I'll leave it to @lukesteensen / @bruceg to give a final approval on the topology changes.

prognant commented:
I rebased against master and addressed most reviews. Regarding the doc, I was thinking updating the doc as soon as a component makes use of this new trace type (see this [comment](https://github.com/vectordotdev/vector/pull/10483#discussion_r782607073)).
Regarding VRL, a tradeoff could be not to allow traces to flow through the `remap` transform while keeping VRL condition available on traces.

prognant commented:
For clarity purpose I split the doc into a [separate branch](11021).
Most of the @lukesteensen concerns should be addressed in the latest iteration.

prognant commented:
@lukesteensen I think I addressed all your suggestions, would you mind taking another look?

lukesteensen reviewed (approved):
Sorry for the delayed review @prognant! This looks good to me.

bruceg commented on lib/vector-core/src/config/mod.rs:
```diff
@@ -12,11 +14,23 @@ pub use log_schema::{init_log_schema, log_schema, LogSchema};
+
+impl fmt::Display for DataType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut t = Vec::new();
+        self.contains(DataType::Log).then(|| t.push("Log"));
+        self.contains(DataType::Metric).then(|| t.push("Metric"));
+        self.contains(DataType::Trace).then(|| t.push("Trace"));
+        write!(f, "{}", t.join(","))
```
Could this be done without the extra allocations? Does it matter (ie is it performance sensitive)? Note that the final line can be simplified to `f.write_str(t.join(","))`

prognant replied:
This is not performance sensitive, this was added to get explicit error message when a typing error occur while resolving the topology. I guess it could be rewritten without allocating a vec, but this would probably add a lot of fuss for something out of the hot path.

bruceg commented on lib/vector-core/src/event/mod.rs:
```diff
@@ -79,6 +85,11 @@ impl Event {
         Event::Log(LogEvent::default())
     }
 
+    #[must_use]
+    pub fn new_empty_trace() -> Self {
+        Event::Trace(TraceEvent::default())
+    }
+
```
I think we'd like to discourage the use of an interface like this in favour of `TraceEvent::default().into()`

bruceg commented on lib/vector-core/src/event/proto.rs:
```diff
@@ -41,6 +47,18 @@ impl From<Log> for event::LogEvent {
     }
 }
 
+impl From<Trace> for event::LogEvent {
+    fn from(trace: Trace) -> Self {
+        let fields = trace
+            .fields
+            .into_iter()
+            .filter_map(|(k, v)| decode_value(v).map(|value| (k, value)))
+            .collect::<BTreeMap<_, _>>();
+
+        Self::from(fields)
+    }
+}
```
I'm curious where this is useful without also setting up the `message` field, etc.

prognant replied:
This is needed because of  https://github.com/vectordotdev/vector/blob/c78a1d670b6ace44e2614f6256c273a20d54ed5b/lib/vector-core/src/event/proto.rs#L144 for conversion from protobuf (hence the dump-all-fields-logic).

bruceg reviewed (commented):
Looks mostly good, I  just have a few minor questions.
